File: scripts/validate-tasks.py
import yaml
import jsonschema
import sys

def validate_tasks(file_path="tasks.yml"):
    """Validate task phases against the schema embedded in the YAML file."""
    # ``file_path`` allows running validation on a custom tasks file
    try:
        with open(file_path, 'r') as f:
            data = yaml.safe_load(f)

        if 'jsonschema' not in data:
            print(f"Error: 'jsonschema' key not found in {file_path}")
            sys.exit(1)

        if 'phases' not in data:
            print(f"Error: 'phases' key not found in {file_path}")
            sys.exit(1)

        schema_str = data['jsonschema']
        schema = yaml.safe_load(schema_str)
        tasks = data['phases']

        jsonschema.validate(instance=tasks, schema=schema)
        print(f"Successfully validated {len(tasks)} tasks in {file_path}.")

    except FileNotFoundError:
        print(f"Error: {file_path} not found.")
        sys.exit(1)
    except yaml.YAMLError as e:
        print(f"Error parsing YAML in {file_path}: {e}")
        sys.exit(1)
    except jsonschema.ValidationError as e:
        print(f"Validation error in {file_path}: {e.message}")
        sys.exit(1)
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        sys.exit(1)

if __name__ == "__main__":
    validate_tasks()

File: scripts/validate-tasks.mjs
import fs from 'fs/promises';
import { pathToFileURL } from 'url';
import yaml from 'yaml';
import Ajv from 'ajv';

export async function validateTasks(filePath = 'tasks.yml') {
  try {
    const data = await fs.readFile(filePath, 'utf8');
    const doc = yaml.parse(data);
    if (!doc.jsonschema) {
      console.error(`Error: 'jsonschema' key not found in ${filePath}`);
      process.exitCode = 1;
      return;
    }
    if (!doc.phases) {
      console.error(`Error: 'phases' key not found in ${filePath}`);
      process.exitCode = 1;
      return;
    }
    const schema = yaml.parse(doc.jsonschema);
    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    if (!validate(doc.phases)) {
      console.error(
        `Validation error in ${filePath}: ${ajv.errorsText(validate.errors)}`
      );
      process.exitCode = 1;
      return;
    }
    console.log(
      `Successfully validated ${doc.phases.length} tasks in ${filePath}.`
    );
  } catch (err) {
    console.error(`An unexpected error occurred: ${err.message}`);
    process.exitCode = 1;
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  const argPath = process.argv[2] || 'tasks.yml';
  validateTasks(argPath);
}
